### ISSUE TITLE
HTWK Gym - Shooting

### ISSUE THREAD

**Issue by marcospaulo429** [AKCIT-RL/robocup_t1#19]
In the HTWK Gym repository, we have implemented a walking and kicking movement for Booster T1. With the goal of creating new movements for Korea, we are studying this repository to improve the kicking movement, as currently it's a very simple stationary kick that connects...

**Comment by marcospaulo429** [comment]
# Report 04/05/2026

After analyzing the initial repository, we concluded that it does the following:

- The robot receives information about the ball's location and attempts to kick it at a fixed point on the map. The ball's position is randomized. The robot starts stationary and attempts to kick while stationary.

To improve this kick, we considered the following step-by-step process:

- [x] Randomize the ball's position on the x-axis.

- [x] Randomize the ball's position on the y-axis.

- [x] Initial position of the robot in motion.

- [x] Create a simulator to evaluate our policy, providing fixed targets and assessing the quality of the kick.

The modifications to the simulator related to the 3 topics have already been made. Since we currently don't have an evaluation simulator, we only have videos of the robot learning to kick, but we are not yet certain that it is aiming at the correct location, and we cannot yet see the rewards of the training.

VIDEO/PHOTO OF WANDB

**Comment by marcospaulo429** [comment]
**New environment: `KickingMovement` (`T1/Kicking_Movement`)**
- Added a full new task (`kicking_movement.py`, ~1400 lines) that combines locomotion (Base Walk) and kicking into a single policy. The robot must walk toward the ball and kick it toward a 3D target — a unified end-to-end skill instead of separate policies.
- Added the corresponding YAML config (`Kicking_Movement.yaml`) with 52-obs unified observation layout, 4096 parallel envs, and tuned reward weights blending base-walk and kicking objectives.

**3D kick targeting (Z-axis)**
- Kick target is now a 3D point (x, y, z) instead of 2D. Target elevation (z) is sampled from a configurable range (`kick_target_z_range`).
- Added `z_error_at_kick` metric: angular difference between ball launch elevation and target elevation, logged per-episode and tracked in W&B.
- Ball velocity and target direction computations upgraded from XY-only to full 3D vectors.

**Unified 52-observation layout**
- `Base_Walk` obs buffer extended to 52 dimensions with 5 placeholder slots (ball pos xy, target direction, target z) set to zero during locomotion training. `KickingMovement` fills these slots with real ball/target data — enabling knowledge transfer from pre-trained walking policies.
- New `Base_Walk_Extended.yaml` config reflects this layout.

**Runner improvements**
- runner.py updated to support the new task registration, checkpoint loading, and env-specific config routing.
- __init__.py registers `KickingMovement`.

**Evaluator (evaluate_kick.py)**
- Evaluator script updated to track `z_error_at_kick` alongside angular and lateral errors, and to use the 3D target when monkey-patching `_resample_kick_target`.

**Reward tuning**
- Several reward parameter adjustments to reduce excessive joint force while maintaining kicking power (commits: "parameters to do less force", "best ball acceleration").
- Ball acceleration reward refined to better capture the impulse at contact.

**Comment by marcospaulo429** [comment]
# Report 06/01/2026

## Summary (May 11 → May 30)

### 1. Split into Bica and Chapa variants

`KickingMovement` was split into two specialized subclasses:

- **`Kicking_Movement_Bica`** — toe-kick: contact with foot pitched ~53° downward. Contains all base kicking logic.
- **`Kicking_Movement_Chapa`** — instep kick: subclass of Bica targeting flat-foot contact. Adds new per-episode metrics: `foot_pitch_at_impact`, `flat_pose_rate`, `instep_contact_rate`.

---

### 2. Multi-scenario evaluator (evaluate_kick.py)

Headless evaluation script covering 6 independent scenarios:

| Scenario | What varies |
|----------|------------|
| `angles` | target angle (−45° → +45°) |
| `ball_pos` | ball position in robot frame |
| `robot_yaw` | initial robot yaw |
| `ball_vel` | initial ball velocity |
| `distance` | target distance (1 m → 8 m) |
| `disturb_push` | external push impulse during approach |

Primary metrics: aggregate hit rate, mean angular error, fall rate.

---

### 3. Bica policy: selection and deployment

5 training runs of 5,000 iterations were evaluated with evaluate_kick.py. Best candidate selected:

| Metric | Result |
|--------|--------|
| Aggregate hit rate | **66.7%** |
| Mean angular error | **14.3°** |
| Fall rate | **4%** |
| Total attempts evaluated | 360 |

- Exported model: kicking_bikinha.pt (TorchScript, 52→12)
- Deploy script: deploy_kicking_bikinha.py (CLI args: `--ball_x`, `--ball_y`, `--kick_angle_deg`, `--config`, `--net`)

---

### 4. Chapa ablation — V1 to V5

**Problem:** V1–V4 all converged to the same toe-kick attractor even with dedicated pose-shaping rewards. Root cause: the dominant `ball_velocity_target_direction` reward (~30 pts/s, sustained ~2 s post-impact) is collected equally regardless of foot pose, so the policy maximized swing speed (= toe whip) and ignored the pose bonuses.

**Solution (V5):** A multiplicative gate on the dominant post-kick reward conditioned on foot pitch at impact (`chapa_velocity_pose_weight`). Results at 5,000 iterations:

| Metric | Bica baseline | V5 (Chapa) |
|--------|:-------------:|:-----------:|
| Foot pitch at impact | 0.92 rad (53°) | **0.058 rad (3.3°)** |
| `flat_pose_rate` (< 0.10 rad) | 3.4% | **82.4%** |
| `instep_contact_rate` | 1.6% | 9.6% |
| `success_rate_20deg` | ~95% | **95.2%** |
| `ball_velocity_target_direction` | ~30 | 6.77 |

**Conclusion:** Foot pose was successfully reshaped (flat foot in 82% of episodes, pitch of ~3°), and accuracy was preserved. However, ball exit velocity collapsed (~77% drop). Analysis shows contact occurs ~19 cm ahead of the foot frame origin — at the leading edge of the horizontal foot rather than the instep (~8 cm). A contact-position shaping mechanism would be needed for the next iteration. Model shipped as kicking_chapa.pt.

---

### 5. MuJoCo interactive evaluator (play_mujoco_kick.py)

Visualization and evaluation script in MuJoCo with two modes:

- **Interactive viewer:** keyboard controls — `R` reset, `B` respawn ball, `[`/`]` rotate target, `-`/`=` change target distance, `P` pause.
- **Headless (`--evaluate`):** mirrors all evaluate_kick.py scenarios in MuJoCo, producing `attempts.csv`, `summary.json`, and `all_scenarios_summary.json` per scenario.

---

### 6. New observation slot: `target_distance` (52 → 53 obs)

Added a target distance slot (ball → target) to all kicking policy observations (`Bica`, `Chapa`, `Kicking`). Distance is now randomized per episode from a configurable range (`kick_target_ref_distance_range: [5.0, 10.0]`) instead of being fixed.

Updated layout (53 obs for Bica / Base Walk Extended):
```
0:3   gravity | 3:6 ang_vel | 6:9 commands | 9:11 gait
11:13 ball_xy | 13:15 target_dir | 15:16 target_z | 16:17 target_distance
17:29 dof_pos | 29:41 dof_vel | 41:53 actions
```

---

### 7. Full real-robot deployment pipeline

TorchScript models and deploy scripts ready for all policies:

| Policy | Script | Obs | Model |
|--------|--------|:---:|-------|
| Base Walk | `deploy_base_walk.py` | 47 | `base_walk.pt` |
| **Base Walk Extended** | `deploy_base_walk_extended.py` | **53** | `base_walk_extended.pt` |
| Parameter Walk | `deploy_parameter_walk.py` | 44 | `parameter_walk.pt` |
| **Kicking** | `deploy_kicking.py` | **48** | `kicking.pt` |
| **Kicking Bikinha** | `deploy_kicking_bikinha.py` | **53** | `kicking_bikinha.pt` |

Official training runs committed to the repo:
- `Base_Walk_Extended`: `logs/T1/.../2026-05-27-22-04-37/` — 5,000 epochs, 53 obs
- `Kicking`: `logs/T1/.../2026-05-27-22-44-57/` — 5,000 epochs, 48 obs
- `Kicking_Movement_Bica`: `logs/T1/.../2026-05-28-13-47-27/` — 5,000 epochs, 53 obs

**Comment by marcospaulo429** [comment]
https://github.com/user-attachments/assets/f71fa3e5-6366-42d7-baf5-80ded5bcfd19

**Comment by gpettro** [comment]
# Technical Documentation: HTWK Kick - Bikinha (Booster T1)

## 1. Problem Diagnosis (Jittering) and New Approach
After the first deployments of the `htwk-gym` framework for Reinforcement Learning (RL) kicks on the Booster T1, a critical level of **jittering** (high-frequency vibration/oscillation) was identified in the joints. 

**Possible Cause:** Incompatibility between the gains ($K_p$ and $K_d$) idealized in the code and the dynamic response of the robot. To solve this issue, a **retraining** of the `htwk_base_walk` (extended) and `htwk_kicking_bikinha` policies was performed using the PD controller parameters that were already validated in the `beyond_mimic` training environment (`booster_train`).

---

## 2. Training Environment Limitations (Isaac Gym vs. Actuator) for New Actuator Values:
* The HTWK framework uses **Isaac Gym**, an environment in which **it was not possible to train using the actuator model**.
* Without the actuator model, the policy is trained under a perfectly idealized joint response (without dynamic losses, delays, or actual motor compliance).
* This requires the deployment PD controller to absorb and mitigate physical imperfections not mapped during training, making the tuning of $K_p$ and $K_d$ in the validation simulator (MuJoCo) and on the real robot even more critical.

---

## 3. Gain Adjustment and XML Correction (T1_23dof.xml)

So that the MuJoCo simulation would faithfully reflect the forces required to stabilize the policy after retraining, the `booster_assets/robots/T1/T1_23dof.xml` asset file was updated with the new `damping` values:

| Joint | Old Damping | New Damping (Trained $K_d$) | Equivalent $K_p$ (Python) |
| :--- | :---: | :---: | :---: |
| Hip Pitch (L/R) | 6.59 | **13.0** | 207.0 |
| Hip Roll (L/R) | 6.01 | **12.0** | 189.0 |
| Hip Yaw (L/R) | 6.01 | **12.0** | 189.0 |
| Knee Pitch (L/R) | 9.59 | **16.0** | 251.0 |
| Ankle Pitch (L/R) | 6.83 | **17.0** | 268.0 |
| Ankle Roll (L/R) | 6.83 | **17.0** | 268.0 |

> **Warning:** This change to the XML affects **all** T1 tasks executed in MuJoCo. If any legacy task strictly depends on the old values, a specific MJCF file should be isolated.

---

## 4. Validation Results in MuJoCo and Deployment
By unifying the gain parameters from `booster_train` and correcting the passive damping in the XML file, the high-frequency oscillations (*jittering*) were slightly mitigated in the simulation.

* **Video Results:**
  * Base Walk Extend: 
  
  https://github.com/user-attachments/assets/cd8e612e-366a-4e7b-8278-4da5cc215488

  * Bikinha: 
  
   https://github.com/user-attachments/assets/08db4bae-e70f-4db5-aaf4-7b6aa0d13436
  
  *  Bikinha MuJoCo: 
  
   https://github.com/user-attachments/assets/f6e2fd55-7778-4875-85b5-91071e23df2a

* **Updated Policies:** The binary files of the RL networks (`.pt` for booster deploy) generated after retraining with the new parameters are saved in the deployment folder on [Drive](https://drive.google.com/drive/folders/1QQUiP6a7KG6RisihYNswnsvBqOdW_P05?usp=drive_link)

---

## 5. Final Remarks
No deployment tests were performed on the real robot, nor was a second attempt made to tune the motor parameters for the hardware. This is because this was already a known issue in another research area of the project, and the robot was traveling at the time.
